### PR TITLE
release: add script to bump repositories

### PR DESCRIPTION
**Pull request by jcvenegas**
Add script that will help to bump versions for all the projects.

Fixes: #49

### CODE REVIEW

**Comment by jcvenegas** [comment]
@klynnrif  could you review  docs changes from this PR.

**Comment by jcvenegas** [comment]
@egernst ping more release automation \o/

File: release/update-repository-version.sh [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+while getopts "hp" opt
+do
+	case $opt in
+		h)	usage 0 ;;
+		p)	PUSH="true" ;;
+	esac
+done
+
+shift $(( $OPTIND - 1 ))
+
```
Comment (jcvenegas) [inline release/update-repository-version.sh]: extra space

File: release/update-repository-version.sh [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+repo=${1:-}
+new_version=${2:-}
+[ -n "${repo}" ] || ( echo "ERROR: repository not provided" && usage 1) 
+[ -n "$new_version" ] || ( echo "ERROR: no new version" && usage 1 )
```
Comment (jcvenegas) [inline release/update-repository-version.sh]: space not needed after echo

File: release/update-repository-version.sh [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+$(get_changes "$current_version")
+
+EOT
+	fi
+
+	info "Updating VERSION file"
+	echo "${new_version}" > VERSION
+	if [ -f "configure.ac" ] ; then
+		sed -i -e \
+			"s|\(AC_INIT(\[[-a-zA-Z\_]*\],\s*\[\)${current_version}\(\]\)|\1${new_version}\2|g"\
```
Comment (jcvenegas) [inline release/update-repository-version.sh]: we dont have autoconf based projects should we keep it?

File: release/update-repository-version.sh [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	# find if project has a release notes script
+	release_notes_script="./scripts/release_notes.sh"
+	notes_file=notes.md
+	if [[ -x "${release_notes_script}" ]]
+	then
+		${release_notes_script} "${current_version}" "${new_version}" > "${notes_file}"
+	else
+		cat << EOT > "${notes_file}"
+# Kata Containers ${new_version}
+
```
Comment (jcvenegas) [inline release/update-repository-version.sh]: double space

File: release/update-repository-version.sh [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+#
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+readonly script_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+readonly script_name="$(basename "${BASH_SOURCE[0]}")"
+
+readonly tmp_dir=$(mktemp -t -d cc-bump.XXXX)
```
Comment (jcvenegas) [inline release/update-repository-version.sh]: cc replace with kata or project-bump

File: release/README.md [hunk 6]
```diff
@@ -2,8 +2,37 @@
...
+projects version is updated to keep the same version in all of them.
+
+To update all the projects versions use:
+
+```bash
+$ make bump-kata-version NEW_VERSION=<new-version>
+```
+
+The makefile target `bump-kata-version` will create a Gihub pull request
+in the kata repositories. The all pull requests will be tested by the
```
Comment (jodh-intel) [inline release/README.md]: s/The all/These/ ?

File: release/README.md [hunk 7]
```diff
@@ -2,8 +2,37 @@
...
+Kata CI to ensure all the project is working before the release. Also,
+the PR will be approved and merged by Kata Containers members.
+
 ## tag_repos.sh ##
 
+After all the Kata repositories are updated with a new version it is needed
+to tag the repositires.
```
Comment (jodh-intel) [inline release/README.md]: Typo: "repositories". I'd reword slightly as:

> After all the Kata repositories are updated with a new version, they need to be tagged.

File: release/Makefile [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+NEW_VERSION := 
+
+REPOS :=
+
+REPOS += runtime
+REPOS += proxy
+REPOS += shim
+REPOS += agent
+REPOS += ksm-throttler
+REPOS += osbuilder
```
Comment (jodh-intel) [inline release/Makefile]: This repo isn't listed in `./release/tag_repos.sh`, but probably should be?

File: release/update-repository-version.sh (hunk 3 above)
Comment (jodh-intel) [reply release/update-repository-version.sh]: I'd remove it for now. But we may need to re-add for https://github.com/kata-containers/shim/issues/83.

/cc @sboeuf.

**Comment by jodh-intel** [comment]
Hi @klynnrif, @intelkevinputnam - ptal.

File: release/README.md [hunk 9]
```diff
@@ -2,8 +2,37 @@
 
 This directory contains tools for Kata Containers releases.
 
+## update-repository-version.sh ##
+
+This script is used to create a Github pull request (a.k.a PR) to change version in all
```
Comment (klynnrif) [inline release/README.md]: Suggested rewrite: This script creates a GitHub pull request (a.k.a PR) to change the version in all

File: release/README.md [hunk 10]
```diff
@@ -2,8 +2,37 @@
...
+the Kata repositories.
+
+To get more information about how to use it , run:
```
Comment (klynnrif) [inline release/README.md]: For more information on using the script, run the following:

File: release/README.md [hunk 11]
```diff
@@ -2,8 +2,37 @@
...
+
+```bash
+$ ./update-repository-version.sh -h
+```
+
+### Update Kata projects to a new version ###
+Kata Containers is devided in multiples projects. On each release all the
```
Comment (klynnrif) [inline release/README.md]: typo "devided" to "divided" + lines 17-18 suggested rewrite: 
Kata Containers is divided into multiple projects. With each release, all project versions are updated to keep the version consistent. 
or 
Kata Containers is divided into multiple projects. With each release, all versions for all projects are updated to keep the versions consistent.

File: release/README.md [hunk 12]
```diff
@@ -2,8 +2,37 @@
...
+projects version is updated to keep the same version in all of them.
+
+To update all the projects versions use:
```
Comment (klynnrif) [inline release/README.md]: To update all versions for all projects, use the following:

File: release/README.md [hunk 13]
```diff
@@ -2,8 +2,37 @@
...
+
+```bash
+$ make bump-kata-version NEW_VERSION=<new-version>
+```
+
+The makefile target `bump-kata-version` will create a Gihub pull request
```
Comment (klynnrif) [inline release/README.md]: Lines 26-29 suggested rewrite: The makefile target `bump-kata-version` creates a GitHub pull request in the kata repositories. These pull requests are tested by the Kata CI to ensure the entire project is working prior to the release. Next, the PR is approved and merged by Kata Containers members.

File: release/README.md (hunk 7 above)
Comment (klynnrif) [reply release/README.md]: I agree with @jodh-intel on this rewrite.

**Comment by jcvenegas** [comment]
@jodh-intel @klynnrif  PR updated.

- Fixed docs
- Added tests for new script

File: release/update-repository-version.sh [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+	popd >> /dev/null
+}
+
+usage(){
+	exit_code="$1"
+	cat <<EOT
+Usage:
+	${script_name} [options] <args>
+Args:
+	<repositoy-name> : Name of repository to fork and send PR from github.com/${organization}
```
Comment (jodh-intel) [inline release/update-repository-version.sh]: Nit: typo - "repository".

File: release/update-repository-version.sh [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+	changes=$(git log --oneline "${current_version}..HEAD") || die "failed to get logs"
+	if [ "${changes}" == "" ]; then
+		echo "Version bump no changes"
+		return
+	fi
+
+	# list all PRs merged from $current_version to HEAD
+	git log --merges "${current_version}..HEAD" | awk '/Merge pull/{getline; getline;print }' | while read pr
+	do
+
```
Comment (jodh-intel) [inline release/update-repository-version.sh]: Nit: blank line.

File: release/update-repository-version.sh [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+	local new_version=$1
+	local current_version=$2
+
+	[ -n "$new_version" ] || die "no new version"
+	[ -n "$current_version" ] || die "no current version"
+
+	printf "release: Kata Containers %s\n\n" "${new_version}"
+
+	get_changes "$current_version"
+
```
Comment (jodh-intel) [inline release/update-repository-version.sh]: Nit: blank line.

File: release/update-repository-version.sh [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+	pushd "${repo}" >> /dev/null
+
+	# All repos we build should have a VERSION file
+	[ -f "VERSION" ] || die "VERSION file not found "
+	current_version="$(cat ./VERSION | grep -v '#')"
+
+	info "Creating PR message"
+	# find if project has a release notes script
+	release_notes_script="./scripts/release_notes.sh"
```
Comment (jodh-intel) [inline release/update-repository-version.sh]: Might be worth putting this at the top of the file to make it stand out more.

File: release/update-repository-version_test.sh [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+OK(){
+	echo "ok"
+}
+output_should_contain(){
+	local output="$1"
+	local text_to_find="$2"
+	[ -n  "$output" ]
+	[ -n  "$text_to_find" ]
+	_=$(echo "${output}" | grep "${text_to_find}")
```
Comment (jodh-intel) [inline release/update-repository-version_test.sh]: This looks a bit magic :)

Could you just say the following?

```
echo "${output}" | grep -q "${text_to_find}"
```

File: release/update-repository-version.sh [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+readonly script_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+readonly script_name="$(basename "${BASH_SOURCE[0]}")"
+
+readonly tmp_dir=$(mktemp -t -d pr-bump.XXXX)
+readonly hub_bin="${tmp_dir}/hub-bin"
```
Comment (jodh-intel) [inline release/update-repository-version.sh]: Does this work with a 2FA key btw?

**Comment by jodh-intel** [comment]
Test is failing:

```
Non existing repo should fail
/tmp/jenkins/workspace/kata-containers-packaging-ubuntu-16-04-PR/go/src/github.com/kata-containers/packaging/.ci/../Makefile:15: recipe for target 'test-release-tools' failed
make[1]: *** [test-release-tools] Error 1
make[1]: Leaving directory '/tmp/jenkins/workspace/kata-containers-packaging-ubuntu-16-04-PR/go/src/github.com/kata-containers/packaging'
Makefile:12: recipe for target 'test' failed
make: *** [test] Error 2
```
